extension/agenthealth/handler/useragent/useragent.go

```diff
@@ -0,0 +1,193 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT
+
+package useragent
+
+import (
+	"fmt"
+	"os"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/google/uuid"
+	telegraf "github.com/influxdata/telegraf/config"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awsemfexporter"
+	"go.opentelemetry.io/collector/otelcol"
+	"go.uber.org/atomic"
+	"golang.org/x/exp/maps"
+
+	"github.com/aws/amazon-cloudwatch-agent/cfg/envconfig"
+	"github.com/aws/amazon-cloudwatch-agent/internal/util/collections"
+	"github.com/aws/amazon-cloudwatch-agent/internal/version"
+	"github.com/aws/amazon-cloudwatch-agent/receiver/adapter"
+)
+
+const (
+	handlerID          = "cloudwatchagent.UserAgentHandler"
+	headerKeyUserAgent = "User-Agent"
+
+	flagRunAsUser                 = "run_as_user"
+	flagContainerInsights         = "container_insights"
+	flagPulse                     = "pulse"
+	flagEnhancedContainerInsights = "enhanced_container_insights"
+	separator                     = " "
+
+	typeInputs     = "inputs"
+	typeProcessors = "processors"
+	typeOutputs    = "outputs"
+)
+
+var (
+	singleton UserAgent
+	once      sync.Once
+)
+
+type UserAgent interface {
+	SetComponents(otelCfg *otelcol.Config, telegrafCfg *telegraf.Config)
+	SetContainerInsightsFlag()
+	Header(isUsageDataEnabled bool) string
+	Listen(listener func())
+}
+
+type userAgent struct {
+	dataLock sync.Mutex
+	id       string
+
+	listenerLock sync.Mutex
+	listeners    []func()
+	isRoot       bool
+
+	inputs     collections.Set[string]
+	processors collections.Set[string]
+	outputs    collections.Set[string]
+
+	inputsStr     atomic.String
+	processorsStr atomic.String
+	outputsStr    atomic.String
+}
+
+var _ UserAgent = (*userAgent)(nil)
+
+func (ua *userAgent) SetComponents(otelCfg *otelcol.Config, telegrafCfg *telegraf.Config) {
+	ua.dataLock.Lock()
+	defer ua.dataLock.Unlock()
+	for _, input := range telegrafCfg.Inputs {
+		ua.inputs.Add(input.Config.Name)
+	}
+	for _, output := range telegrafCfg.Outputs {
+		ua.outputs.Add(output.Config.Name)
+	}
+
+	for _, pipeline := range otelCfg.Service.Pipelines {
+		for _, receiver := range pipeline.Receivers {
+			// trim the adapter prefix from adapted Telegraf plugins
+			name := strings.TrimPrefix(string(receiver.Type()), adapter.TelegrafPrefix)
+			ua.inputs.Add(name)
+		}
+		for _, processor := range pipeline.Processors {
+			ua.processors.Add(string(processor.Type()))
+		}
+		for _, exporter := range pipeline.Exporters {
+			ua.outputs.Add(string(exporter.Type()))
+			if exporter.Type() == "awsemf" {
+				cfg := otelCfg.Exporters[exporter].(*awsemfexporter.Config)
+				if cfg.IsPulseApmEnabled() {
+					ua.outputs.Add(flagPulse)
+				}
+				if cfg.IsEnhancedContainerInsights() {
+					ua.outputs.Add(flagEnhancedContainerInsights)
+				}
+			}
+		}
+	}
+
+	if !ua.isRoot {
+		ua.inputs.Add(flagRunAsUser)
```

Comment:
interesting, are we trying to see how many customers run as non root?

Reply:
This is an existing flag we set in the user agent.